Working with Bounded Contexts and Vertical Slices in Hub (#3)
prenda-school/prenda-school.github.io: docs/Practices/hub/hub.md

# Working with Bounded Contexts in Hub

## Intro
This document defines a number of guidelines to follow while working in Hub. Guidelines for working across projects can be found here (*not yet written*).

These guidelines are intended to lead an engineer towards achieving a set of goals outlined below. These guidelines are based around a set of [useful abstractions](../../glossary.md#useful-abstraction) (UAs).

So what are some of the useful abstractions that we utilize? Some of the notable UAs are Domain Driven Design, Bounded Contexts, Command Query Separation, Verticle Slices Architecture, Functional styled Procedural Programming, Ports and Adapters, The Repository Pattern, and Onion (aka Clean aka Hexagonal) Architecture. Not all UAs are implemented dogmatically, and doing so would lead to many instances of conflict between the UAs. Rather choice core parts from different UAs have been chosen and applied in such a way that these parts are orthogonal to each other and support well the defined goals.

> This document will not attempt to answer all critiques of a given guideline or UA nor will it discuss in depth the tradeoffs of why these guidelines the associated UA(s) are the right tools.

Reviewer: ```suggestion
This document will not attempt to answer all critiques of a given guideline or UA nor will it discuss in depth the tradeoffs of why these guidelines and the associated UA(s) are the right tools.
```
```suggestion
This document will not attempt to answer all critiques of a given guideline or UA nor will it discuss in depth the tradeoffs of why these guidelines the associated UA(s) are the right tools.
```

For more on different UAs check out the [learning doc](learning.md).

## Goals
- Code for a given capability is easy to find.
- Changes to a given behavior are less risky.
- The effects of a change to a domain model shape are clear and should cascade and only to those dependents that actually utilize the changes.
- The API of code implementing a behavior is clear. *You should know exactly what data is required to successfully initiate a behavior as well as understanding the expected failures of that behavior*
- Failures due to essential complexity should be uniquely and easily identifiable and bad data states caused by such errors should be easily remediated.
- Code should be covered by automated tests. *Testing strategy and acceptable coverage are not currently specified, but shoot for as much as possible.*

> - Flexibility for implementation within defined bounraries. *Unlimited flexibility (autonomy) leads to multiple competing standards, analysis paralysis, lack of direction, and a lack of predictability in code. Inflexibility (too many conventions/only a single way to do something) leads to brittle code, overly complicated code for the purpose of adhering to defined rules, difficulty with corner cases. The sweet spot is one where there are multiple correct implementations that each can be derived from a set of UAs or guidelines. This provides confidence to an engineer because they are able to orient on a UA or guideline for direction while having to autonomy to being able to chart the appropriate course to get there.*

Reviewer: ```suggestion
- Flexibility for implementation within defined boundaries. *Unlimited flexibility (autonomy) leads to multiple competing standards, analysis paralysis, lack of direction, and a lack of predictability in code. Inflexibility (too many conventions/only a single way to do something) leads to brittle code, overly complicated code for the purpose of adhering to defined rules, difficulty with corner cases. The sweet spot is one where there are multiple correct implementations that each can be derived from a set of UAs or guidelines. This provides confidence to an engineer because they are able to orient on a UA or guideline for direction while having to autonomy to being able to chart the appropriate course to get there.*
```
```suggestion
- Flexibility for implementation within defined bounraries. *Unlimited flexibility (autonomy) leads to multiple competing standards, analysis paralysis, lack of direction, and a lack of predictability in code. Inflexibility (too many conventions/only a single way to do something) leads to brittle code, overly complicated code for the purpose of adhering to defined rules, difficulty with corner cases. The sweet spot is one where there are multiple correct implementations that each can be derived from a set of UAs or guidelines. This provides confidence to an engineer because they are able to orient on a UA or guideline for direction while having to autonomy to being able to chart the appropriate course to get there.*
```

Reviewer: "while having to autonomy to being able to chart the appropriate course"

Something doesn't seem right in this sentence. Do you mean while having "the autonomy"?

## Non Goals
- Don't Repeat Yourself (DRY) and Brevity at the expensive of clarity. *DRYness and Brevity are often associated with readability and understandability, but these should be seen as side effects of Clarity. If ever in doubt about a decision between dry/brevity and clarity, err on the side of clarity. Flexibility is also valuable given Prenda's current state - history of many pivots and a future requiring pivoting and adaptability.*

## Guidelines
- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Other than the consumption of a command or query, code sharing between bounded contexts should be avoided by default. There may be cases where a bounded context is implemented in a different project. In this case no dependency should be had on the code of that project and integration must be done asynchronously (http call, event handler, etc).

> - [UA: Vertical Slices](./learning.md#vertical-slices), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) The only dependencies between Vertical Slices (different command and queries) should be usages of the command or query function from another slice. Using shared code that sits outside the slice is fine. There are occasional exceptions to this rule, but it is preferred to err on the side of duplication by default. If absolutely need to you find yourself needing to share code between slices it should exist in a `shared` directory at the bouned context root. *Vertical Slices revolve around the axis of change being a slice. We want to couple along the axis of change. Sometimes we might be tempted to share code between two slices. One example is needing to share some type because two slices use a similar shape. An issue arises though when one slice is iterated on and needs to change that shape while another slice doesn't. For this reason it would have been better to duplicate the shape between slices. A rule of thumb is if you plan to share code (type, function, constant, etc.) you need to be VERY sure it represents the exact same concept between slices. Sometimes shared code might actually fit best in an existing shared layer (repository, domain, data)*

Reviewer: ```suggestion
- [UA: Vertical Slices](./learning.md#vertical-slices), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) The only dependencies between Vertical Slices (different command and queries) should be usages of the command or query function from another slice. Using shared code that sits outside the slice is fine. There are occasional exceptions to this rule, but it is preferred to err on the side of duplication by default. If you absolutely need to share code between slices, it should exist in a `shared` directory at the bounded context root. *Vertical Slices revolve around the axis of change being a slice. We want to couple along the axis of change. Sometimes we might be tempted to share code between two slices. One example would be needing to share some type, because two slices use a similar shape. An issue arises though when one slice is iterated on and needs to change that shape while another slice doesn't. For this reason it would have been better to duplicate the shape between slices. A rule of thumb is if you plan to share code (type, function, constant, etc.), you need to be VERY sure it represents the exact same concept between slices. Sometimes shared code might actually fit best in an existing shared layer (repository, domain, data)*
```
```suggestion
- [UA: Vertical Slices](./learning.md#vertical-slices), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) The only dependencies between Vertical Slices (different command and queries) should be usages of the command or query function from another slice. Using shared code that sits outside the slice is fine. There are occasional exceptions to this rule, but it is preferred to err on the side of duplication by default. If absolutely need to you find yourself needing to share code between slices it should exist in a `shared` directory at the bouned context root. *Vertical Slices revolve around the axis of change being a slice. We want to couple along the axis of change. Sometimes we might be tempted to share code between two slices. One example is needing to share some type because two slices use a similar shape. An issue arises though when one slice is iterated on and needs to change that shape while another slice doesn't. For this reason it would have been better to duplicate the shape between slices. A rule of thumb is if you plan to share code (type, function, constant, etc.) you need to be VERY sure it represents the exact same concept between slices. Sometimes shared code might actually fit best in an existing shared layer (repository, domain, data)*
```

- [UA: Errors as Values](./learning.md#errors-as-values), [UA: Domain Driven Design - Domain Modeling](./learning.md#domain-modeling) The known errors of a command/query within a Vertical Slice should be included in its type signature. *It is recommended to coerce exceptions to values for known errors whenever possible, but at minimum no known errors should be exposed from a command/query function by way of thrown exception.*

- [UA: Domain Driven Design - Domain Modeling](./learning.md#domain-modeling) If an error represents a possible bad data state there should be documentation regarding this bad data state and how to remediate it. *At minimum there should be a comment attached to the type definition for a given error that provides appropriate context.*

- [UA: Testing](./learning.md#testing) When a bug is fixed an existing test should be updated to "prove" the fix or a new test should be written to accompany the fix.

- [UA: Validation](./learning.md#validatation) Validation should be performed on all handler request data.

## Remediation of code not meeting guidelines

> There are cases of existing code not meeting these existing guidelines. Refactoring existing code to meet these guidelines should be seen as tech debt and prioritized accordingly. Effort should be taken to have all new code should meet these guidelines. 

Reviewer: ```suggestion
There are cases of existing code not meeting these existing guidelines. Refactoring existing code to meet these guidelines should be seen as tech debt and prioritized accordingly. Effort should be taken to have all new code meet these guidelines. 
```
```suggestion
There are cases of existing code not meeting these existing guidelines. Refactoring existing code to meet these guidelines should be seen as tech debt and prioritized accordingly. Effort should be taken to have all new code should meet these guidelines. 
```

> Any code that must break a guideline should be called out, addressed, and either documented for cases which make sense to break a guideline, or have an associated tech debt ticket created in cases where breaking a guideline is required to quickly remediate a critical issue that following a guideline would delay. *An example of this might be a critical bug that needs to be out ASAP. Maybe a fix was written and through existing automated tests + some manual testing + developer intuiton it is determined that this fix works and won't make the problem worse. In this case it is pragmatic to get this fix out even if it ignores the guideline for testing to prove a fix.*

Reviewer 2: ```suggestion
Any code that must break a guideline should be called out, addressed, and either documented for cases which make sense to break a guideline, or have an associated tech debt ticket created in cases where breaking a guideline is required to quickly remediate a critical issue that following a guideline would delay. *An example of this might be a critical bug that needs to be out ASAP. Maybe a fix was written and through existing automated tests + some manual testing + developer intuition it is determined that this fix works and won't make the problem worse. In this case it is pragmatic to get this fix out even if it ignores the guideline for testing to prove a fix.*
```

Additionally guidelines might be library, framework, or project specific. Defining rules around how to interact with a library for example might be pragmatic if that library has capabilities that by some usages might conflict with existing guidelines.

For the purpose of this document, conventions should be seen as more specific guidelines or extensions to the guidelines.

Naming:
- The entrypoint function for a command or query should start with `command` or `query`.
- The entrypoint function for a command or query should be in a file called `command.ts` or `query.ts` OR in a region called `Command` or `Query`.
- Domain code should either be in files named appropriate to the model/process/rule/etc. within a `domain` directory, or be in a `domain.ts` file, or be in a region called `Domain`.
- Legacy request handlers that are related to a bounded context and are expected to be maintained/iterated should be put in a `legacy-handlers` directory within the appropriate bounded context.
- Integration code for 3rd party services should be in a `Data` directory.

> - Repository code (code that interects with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.

Reviewer 2: ```suggestion
- Repository code (code that intersects with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.
```

Reviewer: interacts

- CRUD type functions that create new records in a single tab, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.

Tests:
- Test files should sit close to the code they test. *Test files that test code in disparate files should be broken up into multiple test files itself.*

SQL:
- Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to it's consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*

Routing:
- Handlers should be wired up to routes in files within a directory called `routes` that is at the bounded context root. **

## The Future

> - In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in a one of the schemas in our postgres instance and utilizing a single shared layer among all consumers within Hub might be the most pragmatic approach.*

Reviewer: ```suggestion
- In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in one of the schemas in our postgres instance and utilizing a single shared layer among all consumers within Hub might be the most pragmatic approach.*
```
```suggestion
- In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in a one of the schemas in our postgres instance and utilizing a single shared layer among all consumers within Hub might be the most pragmatic approach.*
```

> - If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should condsider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bouned context and should house ONLY the data required to support the Vertical Slices of those BCs.*

Reviewer: ```suggestion
- If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should condsider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bounded context and should house ONLY the data required to support the Vertical Slices of those BCs.*
```
```suggestion
- If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should condsider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bouned context and should house ONLY the data required to support the Vertical Slices of those BCs.*
```

> - If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should condsider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bouned context and should house ONLY the data required to support the Vertical Slices of those BCs.*

Reviewer 2: ```suggestion
- If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should consider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bouned context and should house ONLY the data required to support the Vertical Slices of those BCs.*
```

Author: outdated, fixed

> - An potential iteration to reduce coupling between between slices in routing is to have each slice directory be able to define it's own unique route while having a shared layer at the bounded context root then reference those definitions. There are many implementations that could achieve this. One could be to use nested routing where each slice/feature set (we don't utilize feature sets yet) would define it's own router and wire handlers and then the bounded context root would aggregate those routers underneath it's own router. Another option is to have each slice define its own unique path constant and have bounded context root simply reference those constants while wiring. 

Reviewer: ```suggestion
- A potential iteration to reduce coupling between slices in routing is to have each slice directory be able to define it's own unique route while having a shared layer at the bounded context root then reference those definitions. There are many implementations that could achieve this. One could be to use nested routing where each slice/feature set (we don't utilize feature sets yet) would define it's own router and wire handlers and then the bounded context root would aggregate those routers underneath it's own router. Another option is to have each slice define its own unique path constant and have bounded context root simply reference those constants while wiring. 
```

## Recommendations
These recommendations, while value adding, do not necessitate being a guideline and being subject to the remediation requirements of a guideline. They are worth understanding and applying as fit.

- Validate at all IO boundaries (guideline above indicates just handler request data). *Typescript can be a powerful tool, but it is only useful at compile time (as you write and build code). Validating that data being ingested by a process written using Typescript can go a long way in ensuring that the type safety provided by Typescript is reliable. Omitting validation at a boundary leaves room for undesirable behavior and sneaky errors (imagine getting "cannot read property X of undefined" when the type system supposedly doesn't allow it)*

- Endeavour to make invalid states impossible. [Examples](./examples.md#endeavour-to-make-invalid-states-impossible)

## Wiggle Room

> This document doesn't aim to define code styling (commads, linting rules, args as object vs not) nor does it aim to define the implementation of given abstractions as long as they don't deviate from the guidelines. *For example you could have one Vertical Slice utilizing a driver to execute queries directly against a database through a connection while another Vertical Slice utilizes an ORM. As long as both slices don't expose database concerns to the consumer of a given slice then both strategies are acceptable. Another example would be tests. Maybe one slice tests itself by unit tests via isolated dependency and mocks + integration tests for side effecst while another test uses just uses api level integration (acceptance) testing. Both are acceptable as long as there is coverage.*

Reviewer: ```suggestion
This document doesn't aim to define code styling (commands, linting rules, args as object vs not) nor does it aim to define the implementation of given abstractions as long as they don't deviate from the guidelines. *For example you could have one Vertical Slice utilizing a driver to execute queries directly against a database through a connection while another Vertical Slice utilizes an ORM. As long as both slices don't expose database concerns to the consumer of a given slice, then both strategies are acceptable. Another example would be tests. Maybe one slice tests itself by unit tests via isolated dependency and mocks + integration tests for side effects, while another test just uses api level integration (acceptance) testing. Both are acceptable as long as there is coverage.*
```

- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Derived Types](./learning.md#derived-types), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Base types representing the shape of domain models should be in the domain directory at the bounded context root. These types can be shared, often for the purpose of derivatives, among slices.

> - [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Domain types, regardless of whether scoped to a slice or scoped to a bounded context, should not depend on code that is not itself either domain code or is a generic utility type (error handling, ts utils, and such). *This is a concept borrowed from Clean/Hexagonal/Onion architecture and key piece in the internal consistency of our domain model if followed. Exceptions to this rule should be very slim or non existent.*

Reviewer 3: little repeat here.

Author: removed explainer text to reduce repeat.

- [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Repository pattern](./learning.md#repository-pattern), [UA: Persistence Agnosticism](./learning.md#persistence-agnosticism) Repository layer code is the only code that should depend on types representing the shape of database table/view/query/etc. *The implementation details of data persistence should not be exposed to consumers of a domain model that utilizes said persistence. One of the quickest ways to a building a ball of mud is to do let a database schema (slow and risky to change) be the source of truth for the shape of a domain model across all layers and systems. There are cases where this is ok (constrained problem space, prototypes, etc), but this should be avoided by default.*

> - [UA: Slim APIs are Easier to Maintain](./learning.md#slim-apis-are-easier-to-maintain) Don't require more data for a function than is required. *A slim API is less risky to change and easier maintain and reason about.*

Reviewer 3: Is it better for an app to make multiple requests to the API for a create call or one call that does multiple actions.

My example is a `createEnrollment` it needs to do many things for this to succeed. Should those be split into multiple API calls for MicroschoolEnrollment, FundingSourceEnrollment and GradeEnrollment? Or is it better that the App does not need to know about each of those?

Author: Single call. I'd create a command for `createEnrollment` and then have it call the other commands. If you have to cross system boundaries then of course it would be an HTTP call since we don't have any other mechanisms in place for inter-service communication.

Just avoid reaching into a different services database of course.

*I'm not using API here to reference just HTTP APIs. Rather I'm using it generically - it could be the API for a Vertical Slice or module etc. The intended takeaway is if you have to perform an operation that requires properties `x` and `y` from entity `A` that itself has properties `x`, `y`, and `z` - then don't write your endpoint, type signature, or stored procedure etc to take `x`, `y`, AND `z` simply because `z` is a part of entity `A`*

- [UA: Errors as Values](./learning.md#errors-as-values), [UA: Domain Driven Design - Domain Modeling](./learning.md#domain-modeling) The known errors of a command/query within a Vertical Slice should be included in its type signature. *It is recommended to coerce exceptions to values for known errors whenever possible, but at minimum no known errors should be exposed from a command/query function by way of thrown exception.*

- [UA: Domain Driven Design - Domain Modeling](./learning.md#domain-modeling) If an error represents a possible bad data state there should be documentation regarding this bad data state and how to remediate it. *At minimum there should be a comment attached to the type definition for a given error that provides appropriate context.*

- [UA: Testing](./learning.md#testing) When a bug is fixed an existing test should be updated to "prove" the fix or a new test should be written to accompany the fix.

> - [UA: Validation](./learning.md#validatation) Validation should be performed on all handler request data.

Reviewer 3: Meaning we get the data we expect string/int/etc? Or that we are validating the data against the DB?

Author: Validate for:
- Shape *correct fields in correct structure with correct spelling*
- Types *string, number, boolean, etc*
- Invariants *Rules to check for model consistency.* 
  - One example is if property `a` is optional and if `a` is provided then `b` must be provided - but ONLY in the case of `a` being provided.
  - Another example is if `a` has the value `"X"` and `b` must be greater than 600 if `a` is `"X"` - then check for that.
  - Here is a video with an [example](https://youtu.be/64ngP-aUYPc?t=131) *90 sec watch*. Just watch the invariant section of video.
  
*This guideline is more about making sure validation is performed against handler request data. So this is currently just our HTTP request handlers*

Reviewer 4: Would you consider publishing the "Validate for:" section in the doc itself? I found your response to Skylar's question very useful.

Author: Added

## Conventions
Sometimes guidance that provides more specificity than a guideline needs to exist. 

More often than not this is for enabling better implication. For example a naming convention allows engineers to draw better and more accurate conclusions without needing to review documentation.

Additionally guidelines might be library, framework, or project specific. Defining rules around how to interact with a library for example might be pragmatic if that library has capabilities that by some usages might conflict with existing guidelines.

For the purpose of this document, conventions should be seen as more specific guidelines or extensions to the guidelines.

Naming:

> - The entrypoint function for a command or query should start with `command` or `query`.

Reviewer 3: What about ending a function name with `command` or `query`?

Author: I'm indifferent to prefix or postfix. Would you be amenable to a followup proposal PR that postfix by amending the rule in this document + another references PR that modifies the function names in Hub?

Reviewer 3: Sure.

- The entrypoint function for a command or query should be in a file called `command.ts` or `query.ts` OR in a region called `Command` or `Query`.
- Domain code should either be in files named appropriate to the model/process/rule/etc. within a `domain` directory, or be in a `domain.ts` file, or be in a region called `Domain`.
- Legacy request handlers that are related to a bounded context and are expected to be maintained/iterated should be put in a `legacy-handlers` directory within the appropriate bounded context.

> - Integration code for 3rd party services should be in a `Data` directory.

Reviewer 3: Why `Data`? Seems like that would indicate DB or internal data structure. Would another name be more appropriate? Maybe `resources`?

Author: Good point.

Additionally `Data` *I meant lowercase `data` whoops* might seem to indicate that we utilize 3rd party services for just data when we might also use it for doing actions.

`resources` works for me.

Would you be amenable to changing this in a different PR that includes a change to this guideline + the corresponding change in Hub? I can own making this PR.

Reviewer 3: Sure I can do a PR to update this.

- Repository code (code that interacts with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.
- (Request or Event) Handlers exposing a command should be in a file called `api.ts`, a region called `API`, or in files called `request.ts` or `event.ts` in a directory called `handlers`.
- Integration code for bounded contexts in a different project should be in a `dependent-bcs` directory with sub directories matching tha appropriate bounded context. *As of Jan 24 2023 this is not being utilized. Defining this early to reduce decision overhead when this occurs.*
- CRUD type functions that create new records in a single tab, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.

*[Hub Directory Structure Example](examples.md#hub-directory-structure)*

Tests:
- Test files should sit close to the code they test. *Test files that test code in disparate files should be broken up into multiple test files itself.*

SQL:

> - Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to it's consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*

Reviewer 3: This [VSCode plugin](https://marketplace.visualstudio.com/items?itemName=frigus02.vscode-sql-tagged-template-literals) is what I found. What else is there?

Author: I'll be honest. It's been a while since I've done this (used SQL in JS/TS syntax highlighting in my daily work).

I used to be able to go to the bottom right of VSCode and change the language rules to be SQL and it would highlight within a string. I just tried this though and it didn't work. Unsure why.

Back at Pluralsight when we were in office, the work stations there had their dev environment set up to do SQL in JS/TS highlighting in vscode. I just never looked into how they did it.

My current process is just taking the SQL and pasting it in a sql client (I use postico).

**Thoughts on this convention**
- While annoying, there are workarounds to the syntax highlighting that a sql file provides.
- The reason we can't put a SQL file right next to each of it's usages - so like right in each slice that needs it - is because massivejs only support a single scripts directory. You can change the location, but it's still a single directory. *I searched for 5 min and couldn't find a way to have multiple scripts dirs. Maybe a more thorough search might turn up something?*
- While it may be pragmatic now and then to bend a convention/guideline/work against a principle because a tool works a certain way, I believe it's best to be extremely selective when you do this to avoid slippery slopes. I think a good rule of thumb is whether it's blocking to be able to get something done. In our case we can still utilize custom queries with massivejs. There's also workaround to get syntax highlighting or a process workaround by developing the query in a sql client. I think efforts in regaining the ergonomics of being able to utilize a SQL file are best done by finding a better tool than massivejs that can utilize SQL files while being able to have them in more than one spot.

Reviewer 3: I tested out the VS code plugin I mentioned and didn't like the way it worked. I'm in favor of a Typed ORM if we can implement it.

Routing:
- Handlers should be wired up to routes in files within a directory called `routes` that is at the bounded context root. **

## The Future

> - In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in one of the schemas in our postgres instance and utilizing a single shared layer among all consumers within Hub might be the most pragmatic approach.*

Reviewer 3: YES, maybe some shapes for basic DB objects. Or if we can use a DB middleware that can returned typed objects (is that even a thing?)

Author: To be very clear in the communication of this point. I'll add a distinction to this text in the document as well as here in comment. I do mean single shared **repository model**.

There's plenty of tools out there that will returned typed objects. I've used TypeOrm for quite a while and I like it. While it does require some configuration, all the options make sense, it's really mature, well maintained, and not overbearing.

The key things I want in a DB tool is:
- Validate during process startup/making a connection that code defined shapes (type/interface/class) are valid representations of their DB table counterparts and throw/return an error if this is not the case.
- Return to you typed objects when interacting with it's API
- Able to execute custom SQL and provide a generic argument for return query shape.

Reviewer 3: I wonder if we could add [Prisma](https://www.prisma.io/typescript) as the DB query/update tool. Maybe down the road we could make it the full ORM. At least it would give us the types, and have the option to run straight SQL too.

Author: Prisma works tool. Tool choice doesn't matter much to me as much as typesafety and flexibility.

- Endeavour to make invalid states impossible. [Examples](./examples.md#endeavour-to-make-invalid-states-impossible)

## Wiggle Room
This document doesn't aim to define code styling (commands, linting rules, args as object vs not) nor does it aim to define the implementation of given abstractions as long as they don't deviate from the guidelines. *For example you could have one Vertical Slice utilizing a driver to execute queries directly against a database through a connection while another Vertical Slice utilizes an ORM. As long as both slices don't expose database concerns to the consumer of a given slice, then both strategies are acceptable. Another example would be tests. Maybe one slice tests itself by unit tests via isolated dependency and mocks + integration tests for side effects, while another test just uses api level integration (acceptance) testing. Both are acceptable as long as there is coverage.*

## Contributing
TODO

### Process to change
TODO

### Tips for changes
Guidelines:

> - Make sure a guideline implements the intent of a UA - whether it's a new or existing one.

Reviewer 3: UA?

Reviewer 3: Maybe link to the Useful Abstraction definition.

Author: linked to def

- Failures due to essential complexity should be uniquely and easily identifiable and bad data states caused by such errors should be easily remediated.
- Code should be covered by automated tests. *Testing strategy and acceptable coverage are not currently specified, but shoot for as much as possible.*
- Flexibility for implementation within defined boundaries. *Unlimited flexibility (autonomy) leads to multiple competing standards, analysis paralysis, lack of direction, and a lack of predictability in code. Inflexibility (too many conventions/only a single way to do something) leads to brittle code, overly complicated code for the purpose of adhering to defined rules, difficulty with corner cases. The sweet spot is one where there are multiple correct implementations that each can be derived from a set of UAs or guidelines. This provides confidence to an engineer because they are able to orient on a UA or guideline for direction while having to autonomy to being able to chart the appropriate course to get there.*

## Non Goals

> - Don't Repeat Yourself (DRY) and Brevity at the expensive of clarity. *DRYness and Brevity are often associated with readability and understandability, but these should be seen as side effects of Clarity. If ever in doubt about a decision between dry/brevity and clarity, err on the side of clarity. Flexibility is also valuable given Prenda's current state - history of many pivots and a future requiring pivoting and adaptability.*

Reviewer 4: Great to call this out explicitly; reinforces that we value flexibility and clarity WHILE adhering to these UAs as much as reasonable.

- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Derived Types](./learning.md#derived-types), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Base types representing the shape of domain models should be in the domain directory at the bounded context root. These types can be shared, often for the purpose of derivatives, among slices.

- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Domain types, regardless of whether scoped to a slice or scoped to a bounded context, should not depend on code that is not itself either domain code or is a generic utility type (error handling, ts utils, and such). *This is a concept borrowed from Clean/Hexagonal/Onion architecture and key piece in the internal consistency of our domain model if followed. Exceptions to this rule should be very slim or non existent.*

> - [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Repository pattern](./learning.md#repository-pattern), [UA: Persistence Agnosticism](./learning.md#persistence-agnosticism) Repository layer code is the only code that should depend on types representing the shape of database table/view/query/etc. *The implementation details of data persistence should not be exposed to consumers of a domain model that utilizes said persistence. One of the quickest ways to a building a ball of mud is to do let a database schema (slow and risky to change) be the source of truth for the shape of a domain model across all layers and systems. There are cases where this is ok (constrained problem space, prototypes, etc), but this should be avoided by default.*

Reviewer 4: ```suggestion
- [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Repository pattern](./learning.md#repository-pattern), [UA: Persistence Agnosticism](./learning.md#persistence-agnosticism) Repository layer code is the only code that should depend on types representing the shape of database table/view/query/etc. *The implementation details of data persistence should not be exposed to consumers of a domain model that utilizes said persistence. One of the quickest ways to building a ball of mud is to let a database schema (slow and risky to change) be the source of truth for the shape of a domain model across all layers and systems. There are cases where this is ok (constrained problem space, prototypes, etc), but this should be avoided by default.*
```

Reviewer 4: Fixing sentence readability.

- [UA: Slim APIs are Easier to Maintain](./learning.md#slim-apis-are-easier-to-maintain) Don't require more data for a function than is required. *A slim API is less risky to change and easier maintain and reason about.*

> - [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Eventual Consistency](./learning.md#eventual-consistency) All data should only have a single source of truth. *It is ok for data to be stored in multiple places as long as the source of truth is always one location. For concepts that exist within multiple domains (students), it is perfectly acceptable for that concept to have it's source of truth spread across data stores as long as no part (field) of any individual meaning is has it's source of truth in more than one location.*

Reviewer 4: ```suggestion
- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Eventual Consistency](./learning.md#eventual-consistency) All data should only have a single source of truth. *It is ok for data to be stored in multiple places as long as the source of truth is always one location. For concepts that exist within multiple domains (students), it is perfectly acceptable for that concept to have its source of truth spread across data stores as long as no part (field) of any individual meaning is has its source of truth in more than one location.*
```

Reviewer 4: Would it be clearer to change the last sentence to:

"... of any individual meaning is stored in more than one location."

For the purpose of this document, conventions should be seen as more specific guidelines or extensions to the guidelines.

Naming:
- The entrypoint function for a command or query should start with `command` or `query`.
- The entrypoint function for a command or query should be in a file called `command.ts` or `query.ts` OR in a region called `Command` or `Query`.
- Domain code should either be in files named appropriate to the model/process/rule/etc. within a `domain` directory, or be in a `domain.ts` file, or be in a region called `Domain`.
- Legacy request handlers that are related to a bounded context and are expected to be maintained/iterated should be put in a `legacy-handlers` directory within the appropriate bounded context.
- Integration code for 3rd party services should be in a `Data` directory.
- Repository code (code that interacts with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.
- (Request or Event) Handlers exposing a command should be in a file called `api.ts`, a region called `API`, or in files called `request.ts` or `event.ts` in a directory called `handlers`.

> - Integration code for bounded contexts in a different project should be in a `dependent-bcs` directory with sub directories matching tha appropriate bounded context. *As of Jan 24 2023 this is not being utilized. Defining this early to reduce decision overhead when this occurs.*

Reviewer 4: ```suggestion
- Integration code for bounded contexts in a different project should be in a `dependent-bcs` directory with sub directories matching the appropriate bounded context. *As of Jan 24 2023 this is not being utilized. Defining this early to reduce decision overhead when this occurs.*
```

Reviewer 4: Just a typo :)

> - CRUD type functions that create new records in a single tab, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.

Reviewer 4: ```suggestion
- CRUD type functions that create new records in a single table, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.
```

Reviewer 4: Just a typo :)

*[Hub Directory Structure Example](examples.md#)*

Tests:
- Test files should sit close to the code they test. *Test files that test code in disparate files should be broken up into multiple test files itself.*

SQL:

> - Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to it's consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*

Reviewer 4: ```suggestion
- Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to its consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*
```

Routing:
- Handlers should be wired up to routes in files within a directory called `routes` that is at the bounded context root. **

## The Future
- In the very near future it might make sense to refactor out shared repository shapes and functions in the Microschool Administration and Support Bounded Context to a layer that is shared between all bounded contexts. *It's a fact that the lion's share of data utilized by Hub is in one of the schemas in our postgres instance and utilizing a single shared repository model among all consumers within Hub might be the most pragmatic approach.*

- If and as we collect more data/it grows in complexity - we likely will need to apply the same decoupling of domains by way of bounded contexts to our databases. *We should condsider breaking up the database. Technical requirements and prudence should drive the implementation of breakup (schemas vs instances, etc). Each new distinct piece should relate to a given bounded context and should house ONLY the data required to support the Vertical Slices of those BCs.*

> - A potential iteration to reduce coupling between slices in routing is to have each slice directory be able to define it's own unique route while having a shared layer at the bounded context root then reference those definitions. There are many implementations that could achieve this. One could be to use nested routing where each slice/feature set (we don't utilize feature sets yet) would define it's own router and wire handlers and then the bounded context root would aggregate those routers underneath it's own router. Another option is to have each slice define its own unique path constant and have bounded context root simply reference those constants while wiring. 

Reviewer 4: ```suggestion
- A potential iteration to reduce coupling between slices in routing is to have each slice directory be able to define its own unique route while having a shared layer at the bounded context root then reference those definitions. There are many implementations that could achieve this. One could be to use nested routing where each slice/feature set (we don't utilize feature sets yet) would define its own router and wire handlers and then the bounded context root would aggregate those routers underneath its own router. Another option is to have each slice define its own unique path constant and have bounded context root simply reference those constants while wiring. 
```
